index.js

```diff
@@ -73,7 +73,23 @@ function attachToBody (options, req, reply, next) {
       mp.destroy(new Error(`${key} is not allowed as field name`))
       return
     }
-    if (body[key] === undefined) {
+
+    const tokens = (
+      options.enableSpecialNotationKeys === true ||
+      options.enableSpecialNotationKeys === 'objectsOnly'
+    )
+      ? getSerializedKeyTokens(key)
+      : []
+
+    if (tokens.length > 0) {
+      const error = validateSerializedKey(tokens, value)
+      if (error) {
+        mp.destroy(new Error(error))
+        return
+      }
+      const segments = getSerializedKeySegments(options, body, tokens, value)
+      setSerializedKey(segments)
```

Comment:
So we create segments and then pass it directly to setSerializedKey. It feels somehow that there is an artificial separation.

Reply:
Even though this algorithm could be simplified, this format separates concerns in distinct steps (tokenization, validation, segment extraction and appending keys to the body).